src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
+		return mono_object_target (scc->objs [0]);
+
+	jobject jobj = (*env)->CallObjectMethod (env, temporary_peers, arraylist_get, temporary_peer_indices [idx]);
+	return jobject_target (jobj);
+}
+
+// Must call this on any AddReferenceTarget returned by scc_target once done with it
+static void release_scc_target (JNIEnv *env, AddReferenceTarget target) {
+	if (!target.is_mono_object)
+		(*env)->DeleteLocalRef (env, target.jobj);
+}
+
+// Add a reference between objects if both are already known to be MonoObjects which are user peers
+static mono_bool
+add_reference_mono_object (JNIEnv *env, MonoObject *obj, MonoObject *reffed_obj)
+{
+	return add_reference (env, mono_object_target (obj), mono_object_target (reffed_obj));
+}
+
+// Clear a local reference which might be NULL // FIXME: Is it safe to just call DeleteLocalRef with NULL?
+static void
+delete_java_lref (JNIEnv *env, jobject jobj)
+{
+	if (jobj)
+		(*env)->DeleteLocalRef (env, jobj);
+}
+
 static void
 gc_prepare_for_java_collection (JNIEnv *env, int num_sccs, MonoGCBridgeSCC **sccs, int num_xrefs, MonoGCBridgeXRef *xrefs)
 {
-	MonoGCBridgeSCC *scc;
-	int i, j, ref_val;
+	/* Some SCCs might have no IGCUserPeers associated with them, so we must create one */
+	jobject arraylist = NULL, gcuserpeer = NULL;
+	jmethodID gcuserpeer_constructor = NULL, arraylist_get = NULL, arraylist_add = NULL;
+	jobject temporary_peers = NULL;     // This is an ArrayList
+	int temporary_peer_count = 0;       // Number of items in temporary_peers
+	int *temporary_peer_indices = NULL; // Map sccs array index -> temporary_peers index
 
-	ref_val = 1;
-	/* add java refs for items on the list which reference each other */
-	for (i = 0; i < num_sccs; i++) {
-		scc = sccs [i];
-
-		MonoJavaGCBridgeInfo    *bridge_info = NULL;
-		/* start at the second item, ref j from j-1 */
-		for (j = 1; j < scc->num_objs; j++) {
-			bridge_info = get_gc_bridge_info_for_object (scc->objs [j-1]);
-			if (bridge_info != NULL && add_reference (env, scc->objs [j-1], bridge_info, scc->objs [j])) {
-				mono.mono_field_set_value (scc->objs [j-1], bridge_info->refs_added, &ref_val);
-			}
-		}
-		/* ref the first from the last */
+	/* Basic setup before looking at xrefs */
+	for (int i = 0; i < num_sccs; i++) {
+		MonoGCBridgeSCC *scc = sccs [i];
+
+		/* Make sure all objects within a single SCC directly or indirectly reference each other */
 		if (scc->num_objs > 1) {
-			bridge_info = get_gc_bridge_info_for_object (scc->objs [scc->num_objs-1]);
-			if (bridge_info != NULL && add_reference (env, scc->objs [scc->num_objs-1], bridge_info, scc->objs [0])) {
-				mono.mono_field_set_value (scc->objs [scc->num_objs-1], bridge_info->refs_added, &ref_val);
+			MonoGCBridgeSCC *first = scc->objs [0];
+			MonoGCBridgeSCC *prev = first;
+
+			/* start at the second item, ref j from j-1 */
+			for (int j = 1; j < scc->num_objs; j++) {
+				MonoGCBridgeSCC *current = scc->objs [j];
+
+				add_reference_mono_object (env, prev, current);
```

Comment:
Notice a pretty bad problem with both the original and patched versions of this line: add_reference can fail, and if that happens the reference protections for the SCC will be incomplete. This could lead to objects being collected which are still in use.